### PR TITLE
fix: Ensure state persisted to `/var/mail-state` retains correct group

### LINKED ISSUE #2942: Freshclam stopped updating after last update
### Miscellaneous first checks

- [X] I checked that all ports are open and not blocked by my ISP / hosting provider.
- [X] I know that SSL errors are likely the result of a wrong setup on the user side and not caused by DMS itself. I'm confident my setup is correct.

### Affected Component(s)

freshclam updates of antivirus database

### What happened and when does this occur?

```Markdown
I received a mail from the clamav system that freshclam had failed
```

### What did you expect to happen?

```Markdown
freshclam retrieving a database update
```

### How do we replicate the issue?

```Markdown
1. I updated the mailserver:
   docker-compose down; docker-compose pull; docker-compose up --build -d
2. 
3.
...
```

### DMS version

v11.3

### What operating system is DMS running on?

Linux

### Which operating system version?

Debian GNU/Linux 11

### What instruction set architecture is DMS running on?

ARM64 / ARM v8 (and newer)

### What container orchestration tool are you using?

Docker

### docker-compose.yml

```yml
version: '3.3'

services:
  mailserver:
    image: docker.io/mailserver/docker-mailserver:latest
    hostname: ${HOSTNAME}
    domainname: ${DOMAINNAME}
    container_name: ${CONTAINER_NAME}
    env_file: mailserver.env

    # To avoid conflicts with yaml base-60 float, DO NOT remove the quotation marks.
    # More information about the mailserver ports:
    # https://docker-mailserver.github.io/docker-mailserver/edge/config/security/understanding-the-ports/
    ports:
      - "25:25"     # SMTP  (explicit TLS => STARTTLS)
      # "143:143"   # IMAP4 (explicit TLS => STARTTLS)  <-- disabled
      - "465:465"   # ESMTP (implicit TLS)
      - "587:587"   # ESMTP (explicit TLS => STARTTLS)
      - "993:993"   # IMAP4 (implicit TLS)
      - "4190:4190" # ManageSieve
    volumes:
      - "./volumes/maildata:/var/mail"
      - "./volumes/mailstate:/var/mail-state"
      - "./volumes/maillogs:/var/log/mail"
      - "/etc/localtime:/etc/localtime:ro"
      - "./config/:/tmp/docker-mailserver/${SELINUX_LABEL}"              # <-- local (re)configuration
      - "/mnt/proxy/certs/:/etc/letsencrypt"                             # <-- SSL certificates
      - "/mnt/proxy/config/dhparams.pem/:/etc/postfix/dhparams.pem:ro"   # <-- TLS crypto file
      - "/mnt/proxy/config/dhparams.pem/:/etc/dovecot/dh.pem:ro"         # <-- same TLS crypto file

    restart: unless-stopped
    stop_grace_period: 1m
    cap_add: [ "NET_ADMIN", "SYS_PTRACE" ]
```

### Relevant log output

```bash
Sun Dec 18 12:00:01 2022 -> --------------------------------------
Sun Dec 18 12:00:01 2022 -> ERROR: Can't create freshclam.dat in /var/lib/clamav
Sun Dec 18 12:00:01 2022 -> Hint: The database directory must be writable for UID 106 or GID 108
Sun Dec 18 12:00:01 2022 -> ERROR: Failed to save freshclam.dat!
Sun Dec 18 12:00:01 2022 -> WARNING: Failed to create a new freshclam.dat!

ad infinitum, ad nauseam
```

### Other relevant information

I repaired it by logging in to the mailserver container and 
`chgrp 108 /var/lib/clamav; chmod g+w /var/lib/clamav.`

But that won't survive a stop and start. I could add a volume to make the changes persistent. But
it probably should be handled in the Dockerfile.

### What level of experience do you have with Docker and mail servers?

- [ ] I am inexperienced with docker
- [X] I am rather experienced with docker
- [ ] I am inexperienced with mail servers
- [X] I am rather experienced with mail servers
- [ ] I am uncomfortable with the CLI
- [X] I am rather comfortable with the CLI

### Code of conduct

- [X] I have read this project's [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md) and I agree
- [X] I have read the [README](https://github.com/docker-mailserver/docker-mailserver/blob/master/README.md) and the [documentation](https://docker-mailserver.github.io/docker-mailserver/edge/) and I searched the [issue tracker](https://github.com/docker-mailserver/docker-mailserver/issues?q=is%3Aissue) but could not find a solution

### Improvements to this form?

_No response_

### PR DESCRIPTION
**Pull request by polarathene**
# Description

This PR covers changes mostly related to `misc-stack.sh`:
- Some minor refactoring of the conditional handling in the symlink loop (see commit for details).
- Ensures `chown -R` is more explicit about not only setting the user, but also the correct group. Postfix also needed a little adjustment there.
- Moves the related functionality for Amavis to be handled with the rest instead of a separate function call.
- Fixes `misc-stack.sh` rspamd ENV condition to add to `FILES` array when enabled, not disabled. This typo was missed [during initial review](https://github.com/docker-mailserver/docker-mailserver/pull/2902).

Likewise we've had a [similar issue with our `Dockerfile` copying over ClamAV database files](https://github.com/docker-mailserver/docker-mailserver/issues/2942#issuecomment-1383512079) from their official image. This prompted reviewing `misc-stack.sh` which should fix the issue for the user.

Fixes #2942 

## References

- [I previously wrote a similar PR adjusting `misc-stack.sh`](https://github.com/docker-mailserver/docker-mailserver/pull/2273), particularly for Postfix. It may be useful reference for any future maintainer stumbling on these kind of changes.
- Contribution of the [original Amavis fix was in Feb 2017](https://github.com/docker-mailserver/docker-mailserver/pull/510). Another contributor [describes the bug in the associated issue](https://github.com/docker-mailserver/docker-mailserver/issues/507#issuecomment-277800277), which is the same bug we've recently had reported with ClamAV I think. Hence being more thorough with the `chown -R` usage in `misc-stack.sh` (_and making it more clear about the purpose via comment_).
- `COPY --link` was [introduced into the `Dockerfile` in the `v11.2` release](https://github.com/docker-mailserver/docker-mailserver/releases/tag/v11.2.0) in [this Sep 2022 PR](https://github.com/docker-mailserver/docker-mailserver/pull/2755).

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change that does improve existing functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by polarathene** [comment]
CI failure:

```
Error: buildx failed with: ERROR: failed to solve: invalid user index: -1
```

<details>
<summary>Full build log</summary>

```
/usr/bin/docker buildx build --cache-from type=local,src=/tmp/.buildx-cache --cache-to type=local,dest=/tmp/.buildx-cache-new,mode=max --iidfile /tmp/docker-build-push-r1dOWM/iidfile --output type=cacheonly --platform linux/amd64 --metadata-file /tmp/docker-build-push-r1dOWM/metadata-file .
#1 [internal] load .dockerignore
#1 transferring context: 59B done
#1 DONE 0.0s

#2 [internal] load build definition from Dockerfile
#2 transferring dockerfile: 12.09kB done
#2 DONE 0.0s

#3 resolve image config for docker.io/docker/dockerfile:1
#3 ...

#4 [auth] docker/dockerfile:pull token for registry-1.docker.io
#4 DONE 0.0s

#3 resolve image config for docker.io/docker/dockerfile:1
#3 DONE 0.5s

#5 docker-image://docker.io/docker/dockerfile:1@sha256:9ba7531bd80fb0a858632727cf7a112fbfd19b17e94c4e84ced81e24ef1a0dbc
#5 resolve docker.io/docker/dockerfile:1@sha256:9ba7531bd80fb0a858632727cf7a112fbfd19b17e94c4e84ced81e24ef1a0dbc 0.0s done
#5 sha256:1328b32c40fca9bcf9d70d8eccb72eb873d1[124](https://github.com/docker-mailserver/docker-mailserver/actions/runs/3945342546/jobs/6752151872#step:7:128)d72dadce04db8badbe7b08546 9.94MB / 9.94MB 0.1s done
#5 extracting sha256:1328b32c40fca9bcf9d70d8eccb72eb873d1124d72dadce04db8badbe7b08546
#5 extracting sha256:1328b32c40fca9bcf9d70d8eccb72eb873d1124d72dadce04db8badbe7b08546 0.2s done
#5 DONE 0.3s

#6 [auth] clamav/clamav:pull token for registry-1.docker.io
#6 DONE 0.0s

#7 [auth] library/debian:pull token for registry-1.docker.io
#7 DONE 0.0s

#8 [internal] load metadata for docker.io/library/debian:11-slim
#8 DONE 0.2s

#9 [internal] load metadata for docker.io/clamav/clamav:latest
#9 DONE 0.2s

#10 [internal] load build context
#10 DONE 0.0s

#11 [stage-base  1/49] FROM docker.io/library/debian:11-slim@sha256:98d3b4b0cee264301eb1354e0b549323af2d0633e1c43375d0b25c01826b6790
#11 resolve docker.io/library/debian:11-slim@sha256:98d3b4b0cee264301eb1354e0b549323af2d0633e1c43375d0b25c01826b6790 0.0s done
#11 DONE 0.0s

#12 importing cache manifest from local:9656101879626942798
#12 DONE 0.0s

#13 FROM docker.io/clamav/clamav:latest@sha256:314c46478306f1bbf3216e2a8ca4b3cb87ba5dd1e14fe4d43f0e3d13712a4af1
#13 resolve docker.io/clamav/clamav:latest@sha256:314c46478306f1bbf3216e2a8ca4b3cb87ba5dd1e14fe4d43f0e3d13712a4af1
#13 ...

#10 [internal] load build context
#10 transferring context: 295.46kB 0.0s done
#10 DONE 0.1s

#13 FROM docker.io/clamav/clamav:latest@sha256:314c46478306f1bbf3216e2a8ca4b3cb87ba5dd1e14fe4d43f0e3d13712a4af1
#13 resolve docker.io/clamav/clamav:latest@sha256:314c46478306f1bbf3216e2a8ca4b3cb87ba5dd1e14fe4d43f0e3d13712a4af1 0.0s done
#13 sha256:24cb207b181c6103680fa838d6c19b2baa1869aee6afcd891ce46b7b174f7b26 1.26kB / 1.26kB 0.0s done
#13 sha256:2d8c7c5ce39eaa53a3997135d3c16598f489f99655024514de17424d064b03e4 348B / 348B 0.0s done
#13 sha256:68331520d622a7054a5e17fa38f2f5e25bf0680d93adfee677445f671ec4e3ba 1.66MB / 1.66MB 0.1s done
#13 sha256:06486e751a263a61f8582b3ef09ed4a80cdc2a3dbce6d1832dd2e0f2f8cf9e33 7.34MB / 231.14MB 0.2s
#13 sha256:06486e751a263a61f8582b3ef09ed4a80cdc2a3dbce6d1832dd2e0f2f8cf9e33 19.92MB / 231.14MB 0.3s
#13 sha256:c158987b05517b6f2c5913f3acef1f2182a32345a304fe357e3ace5fadcad715 3.37MB / 3.37MB 0.1s done
#13 sha256:cff7aab4088b8607efe477bdd4a9e0185bc8ed8339e171980897550e15390466 11.53MB / 78.00MB 0.2s
#13 extracting sha256:c158987b05517b6f2c5913f3acef1f2182a32345a304fe357e3ace5fadcad715
#13 sha256:cff7aab4088b8607efe477bdd4a9e0185bc8ed8339e171980897550e15390466 30.41MB / 78.00MB 0.3s
#13 extracting sha256:c158987b05517b6f2c5913f3acef1f2182a32345a304fe357e3ace5fadcad715 0.2s done
#13 extracting sha256:68331520d622a7054a5e17fa38f2f5e25bf0680d93adfee677445f671ec4e3ba
#13 sha256:06486e751a263a61f8582b3ef09ed4a80cdc2a3dbce6d1832dd2e0f2f8cf9e33 34.83MB / 231.14MB 0.5s
#13 sha256:cff7aab4088b8607efe477bdd4a9e0185bc8ed8339e171980897550e15390466 47.63MB / 78.00MB 0.5s
#13 sha256:06486e751a263a61f8582b3ef09ed4a80cdc2a3dbce6d1832dd2e0f2f8cf9e33 57.67MB / 231.14MB 0.6s
#13 sha256:cff7aab4088b8607efe477bdd4a9e0185bc8ed8339e171980897550e15390466 59.77MB / 78.00MB 0.6s
#13 sha256:cff7aab4088b8607efe477bdd4a9e0185bc8ed8339e171980897550e15390466 77.59MB / 78.00MB 0.8s
#13 sha256:06486e751a263a61f8582b3ef09ed4a80cdc2a3dbce6d1832dd2e0f2f8cf9e33 83.96MB / 231.14MB 0.9s
#13 sha256:cff7aab4088b8607efe477bdd4a9e0185bc8ed8339e171980897550e15390466 78.00MB / 78.00MB 0.9s done
#13 extracting sha256:68331520d622a7054a5e17fa38f2f5e25bf0680d93adfee677445f671ec4e3ba 0.5s done
#13 extracting sha256:cff7aab4088b8607efe477bdd4a9e0185bc8ed8339e171980897550e15390466
#13 sha256:06486e751a263a61f8582b3ef09ed4a80cdc2a3dbce6d1832dd2e0f2f8cf9e33 101.71MB / 231.14MB 1.1s
#13 sha256:06486e751a263a61f8582b3ef09ed4a80cdc2a3dbce6d1832dd2e0f2f8cf9e33 114.29MB / 231.14MB 1.2s
#13 sha256:06486e751a263a61f8582b3ef09ed4a80cdc2a3dbce6d1832dd2e0f2f8cf9e33 [131](https://github.com/docker-mailserver/docker-mailserver/actions/runs/3945342546/jobs/6752151872#step:7:135).76MB / 231.14MB 1.4s
#13 sha256:06486e751a263a61f8582b3ef09ed4a80cdc2a3dbce6d1832dd2e0f2f8cf9e33 152.70MB / 231.14MB 1.5s
#13 sha256:06486e751a263a61f8582b3ef09ed4a80cdc2a3dbce6d1832dd2e0f2f8cf9e33 170.92MB / 231.14MB 1.7s
#13 sha256:06486e751a263a61f8582b3ef09ed4a80cdc2a3dbce6d1832dd2e0f2f8cf9e33 188.74MB / 231.14MB 1.8s
#13 sha256:06486e751a263a61f8582b3ef09ed4a80cdc2a3dbce6d1832dd2e0f2f8cf9e33 214.96MB / 231.14MB 2.0s
#13 sha256:06486e751a263a61f8582b3ef09ed4a80cdc2a3dbce6d1832dd2e0f2f8cf9e33 229.64MB / 231.14MB 2.1s
#13 sha256:06486e751a263a61f8582b3ef09ed4a80cdc2a3dbce6d1832dd2e0f2f8cf9e33 231.14MB / 231.14MB 3.0s done
#13 extracting sha256:cff7aab4088b8607efe477bdd4a9e0185bc8ed8339e171980897550e15390466 2.3s done
#13 extracting sha256:2d8c7c5ce39eaa53a3997[135](https://github.com/docker-mailserver/docker-mailserver/actions/runs/3945342546/jobs/6752151872#step:7:139)d3c16598f489f99655024514de17424d064b03e4 done
#13 extracting sha256:24cb207b181c6103680fa838d6c19b2baa1869aee6afcd891ce46b7b174f7b26 done
#13 DONE 3.3s

#13 FROM docker.io/clamav/clamav:latest@sha256:314c46478306f1bbf3216e2a8ca4b3cb87ba5dd1e14fe4d43f0e3d[137](https://github.com/docker-mailserver/docker-mailserver/actions/runs/3945342546/jobs/6752151872#step:7:141)12a4af1
#13 extracting sha256:06486e751a263a61f8582b3ef09ed4a80cdc2a3dbce6d[183](https://github.com/docker-mailserver/docker-mailserver/actions/runs/3945342546/jobs/6752151872#step:7:187)2dd2e0f2f8cf9e33
#13 extracting sha256:06486e751a263a61f8582b3ef09ed4a80cdc2a3dbce6d1832dd2e0f2f8cf9e33 1.5s done
#13 DONE 4.8s

#14 [stage-base  3/49] RUN chmod +x /usr/local/bin/sedfile
#14 CACHED

#15 [stage-base  4/49] COPY target/scripts/build/* /build/
#15 CACHED

#16 [stage-base  5/49] COPY target/scripts/helpers/log.sh /usr/local/bin/helpers/log.sh
#16 CACHED

#17 [stage-base  6/49] RUN /bin/bash /build/packages.sh
#17 CACHED

#18 [stage-base  2/49] COPY target/bin/sedfile /usr/local/bin/sedfile
#18 CACHED

#19 [stage-base  7/49] RUN <<EOF (echo '0 */6 * * * clamav /usr/bin/freshclam --quiet' >/etc/cron.d/clamav-freshclam...)
#19 CACHED
ERROR: failed to solve: invalid user index: -1
```

</details>

---

This seems to be a bug with the CI caching as I can run `make build` locally fine and it correctly applies the `--chown` change.

**Review by georglauterbach — APPROVED** [review]
While it's never nice to use a workaround, good job at implementing it! LGTM 👍🏼

File: target/scripts/startup/misc-stack.sh [hunk 1]
```diff
@@ -43,36 +44,46 @@ function _misc_save_states
...
+    # can change the values in the Docker image, causing an ownership mismatch.
     _log 'trace' 'Fixing /var/mail-state/* permissions'
-    [[ ${ENABLE_CLAMAV} -eq 1 ]] && chown -R clamav /var/mail-state/lib-clamav
-    [[ ${ENABLE_SPAMASSASSIN} -eq 1 ]] && chown -R debian-spamd /var/mail-state/lib-spamassassin
-    [[ ${ENABLE_POSTGREY} -eq 1 ]] && chown -R postgrey /var/mail-state/lib-postgrey
+    [[ ${ENABLE_AMAVIS}       -eq 1 ]] && chown -R amavis:amavis /var/mail-state/lib-amavis
+    [[ ${ENABLE_CLAMAV}       -eq 1 ]] && chown -R clamav:clamav /var/mail-state/lib-clamav
+    [[ ${ENABLE_FETCHMAIL}    -eq 1 ]] && chown -R fetchmail:nogroup /var/mail-state/lib-fetchmail
+    [[ ${ENABLE_POSTGREY}     -eq 1 ]] && chown -R postgrey:postgrey /var/mail-state/lib-postgrey
+    [[ ${ENABLE_SPAMASSASSIN} -eq 1 ]] && chown -R debian-spamd:debian-spamd /var/mail-state/lib-spamassassin
```
Comment (casperklein) [inline target/scripts/startup/misc-stack.sh:74]:
```suggestion
    [[ ${ENABLE_AMAVIS}       -eq 1 ]] && chown -R amavis:amavis             /var/mail-state/lib-amavis
    [[ ${ENABLE_CLAMAV}       -eq 1 ]] && chown -R clamav:clamav             /var/mail-state/lib-clamav
    [[ ${ENABLE_FETCHMAIL}    -eq 1 ]] && chown -R fetchmail:nogroup         /var/mail-state/lib-fetchmail
    [[ ${ENABLE_POSTGREY}     -eq 1 ]] && chown -R postgrey:postgrey         /var/mail-state/lib-postgrey
    [[ ${ENABLE_SPAMASSASSIN} -eq 1 ]] && chown -R debian-spamd:debian-spamd /var/mail-state/lib-spamassassin
```

I couldn't resist, to propose it 😆

**Review by casperklein — APPROVED** [review]

**Review by casperklein — DISMISSED** [review]

**Review by georglauterbach — APPROVED** [review]

**Comment by polarathene** [comment]
## Users + Groups we add

While working on this PR, I looked into the users we explicitly create or are added from packages installed.

For the majority I've looked up the source for how they were created and documented that all below. Might be useful reference for maintenance someday? :sweat_smile: 

<details>
<summary>We could technically create those users upfront ourselves.</summary>

There isn't much benefit doing that except for when I experimented with splitting different package installs into separate stages and using a naive `COPY --from=some-stage / /` (_as that would overwrite the common `/etc/passwd`, `/etc/group`, `/etc/shadow` files affected_).

That used a fair amount of additional space as `COPY` layers (_even with `--link` if I recall correctly_), but minimized with another separate stage to bundle copies into, and a single `COPY` from that into the output image layers was the same with `--link` (but larger build cache size) and a few MB less without `--link` (_still used extra build cache for the separate stages, which can build concurrently_). Alternatively if each stage depended on the previous instead, it seems to not bulk up the build cache.

One benefit is a central place of where each user is setup, although there's a small chance of maintenance drift from upstream at risk. And should we change these due to add/removing packages, where the UID/GID values shift anyway, then it's not preventing the main problem :sweat_smile: (_hence these `chown -R` startup scripts to patch up such changes_)

</details>

```bash
#!/bin/bash

# -eE :: exit on error (do this in functions as well)
# -u  :: show (and exit) when using unset variables
# -o pipefail :: exit on error in pipes
set -eE -u -o pipefail

function _setup_users() {
  addgroup --quiet --system --force-badname ssl-cert
  addgroup --quiet --system crontab
  addgroup --quiet --system postdrop

  # `--disabled-password` and `--disabled-login` aren't necessary, results in roughly the same /etc/shadow output
  # `--disabled-password` allows for login that is not password-based. It appears to be default for `--system`, retained for clarity.
  local COMMON=(
    --quiet
    --system
    --disabled-password
  )

  # clamav + fetchmail: Should both prefer `nologin` to `/bin/false`
  # for their shell as it is effectively equivalent.
  # fetchmail installs with a home group of nobody, so we add that user without `--group`
  adduser "${COMMON[@]}" --home '/var/lib/fetchmail'    fetchmail

  COMMON+=('--group')
  adduser "${COMMON[@]}" --home '/run/opendkim'         opendkim
  adduser "${COMMON[@]}" --home '/run/opendmarc'        opendmarc
  adduser "${COMMON[@]}" --home '/var/lib/redis'        redis
  # Amavis needs a shell for Dockerfile `razor-admin` lines:
  adduser "${COMMON[@]}" \
    --home '/var/lib/amavis' \
    --gecos 'AMaViS system user' \
    --shell /bin/sh \
    amavis

  COMMON+=('--no-create-home')
  adduser "${COMMON[@]}" --home '/nonexistent'          policyd-spf
  adduser "${COMMON[@]}" --home '/var/spool/postfix'    postfix
  adduser "${COMMON[@]}" --home '/var/lib/postgrey'     postgrey
  adduser "${COMMON[@]}" --home '/var/lib/postsrsd'     postsrsd
  adduser "${COMMON[@]}" --home '/var/lib/spamassassin' debian-spamd

  adduser "${COMMON[@]}" \
    --home '/usr/lib/dovecot' \
    --gecos 'Dovecot mail server' \
    dovecot

  adduser "${COMMON[@]}" \
    --home '/nonexistent' \
    --gecos 'Dovecot login user' \
    dovenull

  adduser "${COMMON[@]}" \
    --home '/var/lib/rspamd' \
    --gecos 'rspamd spam filtering system' \
    --force-badname \
    _rspamd

  # Explicit UID set (--group will use it also for GID), for use with `COPY --chown`
  adduser "${COMMON[@]}" --home '/var/lib/clamav' --uid 200 clamav

  # Add clamav and amavis users to each others group
  adduser clamav amavis
  adduser amavis clamav

  # rsyslog package in Debian (unlike in Ubuntu) does not add the syslog user
  adduser "${COMMON[@]}" --home '/nonexistent' syslog

  # DMS user (used by Dovecot?):
  # NOTE: `adduser` isn't suitable for non-interactive creation of users with passwords.
  # NOTE: This requires openssl to already be installed. A group is implicitly created matching user name and uid values.
  useradd --uid 5000 --home-dir '/home/docker' --shell '/bin/bash' --password "$(echo docker | openssl passwd -1 -stdin)" docker
}

_setup_users

### Additional References ###

# Docs for adduser + useradd:
# https://manpages.debian.org/bullseye/adduser/adduser.8.en.html
# https://manpages.debian.org/bullseye/passwd/useradd.8.en.html

# Default additions to /etc/passwd without this script:
#
# postfix:x:101:102::/var/spool/postfix:/usr/sbin/nologin
# fetchmail:x:102:65534::/var/lib/fetchmail:/bin/false
# postsrsd:x:103:104::/var/lib/postsrsd:/usr/sbin/nologin
# opendkim:x:104:106::/run/opendkim:/usr/sbin/nologin
# opendmarc:x:105:107::/run/opendmarc:/usr/sbin/nologin
# clamav:x:106:108::/var/lib/clamav:/bin/false
# postgrey:x:107:109::/var/lib/postgrey:/usr/sbin/nologin
# policyd-spf:x:108:110::/nonexistent:/usr/sbin/nologin
# debian-spamd:x:109:111::/var/lib/spamassassin:/usr/sbin/nologin
# amavis:x:110:112:AMaViS system user,,,:/var/lib/amavis:/bin/sh
# dovecot:x:111:113:Dovecot mail server,,,:/usr/lib/dovecot:/usr/sbin/nologin
# dovenull:x:112:114:Dovecot login user,,,:/nonexistent:/usr/sbin/nologin
# _rspamd:x:113:115:rspamd spam filtering system,,,:/var/lib/rspamd:/usr/sbin/nologin
# redis:x:114:116::/var/lib/redis:/usr/sbin/nologin
# syslog:x:115:65534::/home/syslog:/usr/sbin/nologin
# docker:x:5000:5000::/home/docker:/bin/bash

# Default additions to /etc/group without this script:
#
# ssl-cert:x:101:
# postfix:x:102:
# postdrop:x:103:
# postsrsd:x:104:
# crontab:x:105:
# opendkim:x:106:
# opendmarc:x:107:
# clamav:x:108:amavis
# postgrey:x:109:
# policyd-spf:x:110:
# debian-spamd:x:111:
# amavis:x:112:clamav
# dovecot:x:113:
# dovenull:x:114:
# _rspamd:x:115:
# redis:x:116:
# docker:x:5000:

# Default additions to /etc/shadow without this script:
# '*' is no password, cannot login as these system users.
# --disabled-login causes the locked '!' value for clamav,spamassassin,rspamd.
#
# postfix:*:19377:0:99999:7:::
# fetchmail:*:19377:0:99999:7:::
# postsrsd:*:19377:0:99999:7:::
# opendkim:*:19377:0:99999:7:::
# opendmarc:*:19377:0:99999:7:::
# clamav:!:19377:0:99999:7:::
# postgrey:*:19377:0:99999:7:::
# policyd-spf:*:19377:0:99999:7:::
# debian-spamd:!:19377:0:99999:7:::
# amavis:*:19377:0:99999:7:::
# dovecot:*:19377:0:99999:7:::
# dovenull:*:19377:0:99999:7:::
# _rspamd:!:19377:0:99999:7:::
# redis:*:19377:0:99999:7:::
# syslog:*:19377:0:99999:7:::
# docker:$1$/ZRgRKw5$kXOPJlnHGZLVzNV8Tfcav/:19377:0:99999:7:::


### Package sources for users and groups created above ###

# amavis (adduser --system --ingroup amavis --home /var/lib/amavis --gecos "AMaViS system user" --shell /bin/sh --quiet --disabled-password amavis)
# https://packages.debian.org/source/bullseye/amavisd-new
# https://salsa.debian.org/debian/amavisd-new/-/blob/dda3c9ef0ec85ac8bf4a3f26caa82e96fbb6d381/debian/amavisd-new.postinst#L53-55

# clamav (adduser --system --no-create-home --quiet --disabled-password --disabled-login --shell /bin/false --group --home /var/lib/clamav clamav)
# https://packages.debian.org/source/bullseye/clamav
# https://salsa.debian.org/clamav-team/clamav/-/blob/4a25d28c0379c4b034d60c686f63f3d518d85286/debian/clamav-base.postinst.in#L35-38

# clamav installs cron package, adds crontab group
# addgroup --system crontab
# https://packages.debian.org/source/bullseye/cron
# https://salsa.debian.org/debian/cron/-/blob/bf78c756347c4533b1eba1ee24aadf6952b03531/debian/cron-daemon-common.postinst#L13

# user: postfix, groups: postfix postdrop
# addgroup --system postdrop
# addgroup --system postfix
# adduser --system --home /var/spool/postfix --no-create-home --disabled-password --ingroup postfix postfix
# https://packages.debian.org/source/bullseye/postfix
# https://salsa.debian.org/postfix-team/postfix-dev/-/blob/ddcd3d9659338f44de47c2e8ece866ea0ade85c5/debian/postfix.postinst#L210-213
# https://salsa.debian.org/postfix-team/postfix-dev/-/blob/ddcd3d9659338f44de47c2e8ece866ea0ade85c5/debian/postfix.postinst#L21-22

# ssl-cert package installed as part of postfix package
# addgroup --quiet --system --force-badname ssl-cert
# https://packages.debian.org/source/bullseye/ssl-cert
# https://salsa.debian.org/apache-team/ssl-cert/-/blob/6d7133f393d044297418a34ca0738276005e55de/debian/postinst#L9

# opendkim (adduser --quiet --system --group --home /run/opendkim opendkim)
# https://packages.debian.org/source/bullseye/opendkim
# https://salsa.debian.org/debian/opendkim/-/blob/3a900f7373dce378c425ca832dea7bfa3e230d3d/debian/opendkim.postinst#L24

# opendmarc (adduser --quiet --system --group --home /run/opendmarc opendmarc)
# https://packages.debian.org/source/bullseye/opendmarc
# https://salsa.debian.org/kitterman/opendmarc/-/blob/b2ba99c5918d0b50048239090f61d985fe970e30/debian/opendmarc.postinst#L28

# postsrsd (adduser --quiet --system --group --no-create-home --home /var/lib/postsrsd postsrsd)
# https://packages.debian.org/source/bullseye/postsrsd
# https://salsa.debian.org/debian/postsrsd/-/blob/9d0c2fbc1b5df858d0fd27763eb4cc2f6f0d9352/debian/postsrsd.postinst#L34

# postgrey
# Git source presently in transition to salsa.debian.org, lacking debian post-install script for postgrey user?
# Installation output does indicate `--no-create-home` used, but prior `/var/lib/postgrey` is created or `chown postgrey:postgrey`.
# https://packages.debian.org/source/bullseye/postgrey
# https://salsa.debian.org/debian/postgrey

# fetchmail (Missing Debian package source)
# https://packages.debian.org/source/bullseye/fetchmail

# policyd-spf (adduser --quiet --system --group --no-create-home --home /nonexistent policyd-spf)
# https://packages.debian.org/source/bullseye/postfix-policyd-spf-perl
# https://salsa.debian.org/kitterman/postfix-policyd-spf-perl/-/blob/38b5b4d4b83f564651261f830e895ee2769ce62e/debian/postfix-policyd-spf-perl.postinst#L7-10

# debian-spamd (adduser --system --group --disabled-login --disabled-password --home /var/lib/spamassassin --no-create-home debian-spamd)
# https://packages.debian.org/source/bullseye/spamassassin
# https://salsa.debian.org/debian/spamassassin/-/blob/ead27ee27f508f0b91f408e688b6446b15cd3018/debian/spamassassin.postinst#L8-10

# dovecot (adduser --system --group --home /usr/lib/dovecot --gecos "Dovecot mail server" --no-create-home --disabled-password --quiet dovecot || true)
# dovenull (adduser --system --group --home /nonexistent --no-create-home --gecos "Dovecot login user" --disabled-password --quiet dovenull || true)
# https://packages.debian.org/source/bullseye/dovecot
# https://salsa.debian.org/debian/dovecot/-/blob/5092256610ddd679e366402abbe7854058f8a78b/debian/dovecot-core.postinst#L35-39

# _rspamd (adduser --quiet --system --group --home /var/lib/rspamd --no-create-home --disabled-login --gecos "rspamd spam filtering system" --force-badname _rspamd)
# https://packages.debian.org/source/bullseye/rspamd
# https://salsa.debian.org/debian/rspamd/-/blob/895f9465cc3559f6f50ba7093d61b57173192f46/debian/postinst#L37-45

# redis (adduser --system --home /var/lib/redis --quiet --group redis || true)
# https://packages.debian.org/source/bullseye/redis
# https://salsa.debian.org/lamby/pkg-redis/-/blob/9f2df070e3a23e5eae432f8e44d3925f1d291830/debian/redis-tools.postinst#L23-28
```

---

## Package install layer weight

I was curious and observed the following breakdown (_`docker image history` layer sizes_):

`build/packages.sh` layer is 300 MB (_up to 326 MB when split into separate stages_) and approx 95 MB compressed (_base image is 80 MB, approx 30MB compressed_).
- `_pre_installation_steps`
- `_install_postfix` (42 MB)
- `_install_packages` (197 MB, ~210 MB if each sub-group is it's own stage)
  - `ANTI_VIRUS_SPAM_PACKAGES` (110 MB)
  - `CODECS_PACKAGES` (12 MB)
  - `MISCELLANEOUS_PACKAGES` (73 MB)
  - `POSTFIX_PACKAGES` (6 MB)
  - `MAIL_PROGRAMS_PACKAGES` (10 MB)
- `_install_dovecot` (32 MB)
- `_install_rspamd` (40 MB)
- `_install_fail2ban` (2 MB)
- `_post_installation_steps`

And then we add the 230MB ClamAV files to `/var/lib/clamav`.

---

## `COPY --link --chown=clamav`

I realized that the `COPY --link` in the separate stage doesn't seem relevant and could instead be `COPY --chown=clamav` if it can source the `clamav` user installed. Then the `RUN chown -R` isn't required, saving 230MB from build cache size.

While `COPY --link` is not compatible with `--chown=clamav` together, we can provide numeric IDs which is compatible. However we'd need to know those values in advance and they should not be prone to change as packages are added / removed in future.

If we create the `clamav` user in advance prior to installing the pacakge, this is no longer a concern and the package install will use that already setup user. Thus we create that user with an explicit UID + GID and we can drop the separate workaround stage, removing another 230MB from build cache.

This keeps the CI build cache that we store back to the same ~345MB size :+1:

**Review by georglauterbach — APPROVED** [review]

**Review by casperklein — DISMISSED** [review]

**Comment by polarathene** [comment]
Err ok that's a first for me. I thought a rebase on master button kept the approval like it does when you update with a merge commit from master :thinking: 

Won't be doing that again then, sorry about that :sweat_smile:

**Review by georglauterbach — APPROVED** [review]

**Review by casperklein — APPROVED** [review]

**Comment by polarathene** [comment]
@georglauterbach are you alright with this branch being merged? Or do you want to wait until your rspamd branch is merged first to avoid dealing with the likely merge conflict on `misc-stack.sh`?

**Comment by georglauterbach** [comment]
Go ahead 🚀

**Pull request merged**